### DISCUSSION TITLE
How to bundle for lambda?

### DISCUSSION THREAD

**Discussion by mattfysh** [nrwl/nx#10970] [General]
One of my packages is the entry point for a lambda, and it consumes other packages. I also have another cloud package which is used to deploy via Pulumi (using the nx-pulumi plugin). The pulumi code references the build assets in the `dist` directory

How can I work with the existing JS/TS plugins and configure a package to bundle, instead of just compiling each file individually? right now I am working around nx by using a manual build step with esbuild to bundle, but I'd prefer to avoid going against the grain, and work with nx's default building blocks as much as possible

Currently when using SWC, I can build the function package into the dist directory, but when I direct pulumi to upload the contents of `dist/my-function` it does not include any local or third party dependencies in the uploaded archive, resulting in failure to resolve imports when running the function

**Comment by narcisonunez** [comment]
Same issue here. Is there any updates for this?

@mattfysh Were you able to find a solution?

**Comment by atanaskanchev** [comment]
Similar issue with bundling Google  cloud functions v2 🤔

**Comment by borgoat** [comment]
I can share our experience since we recently moved away from CDK's own bundling implementation via esbuild[^2], in favour of the Nx esbuild plugin. With caching and parallel execution, this reduced the bundling time from ~1 minute (for ~50 functions), to 10 seconds at most. (There's probably still room for improvement, since invoking plain esbuild with a glob pattern takes between 1/2 seconds, but again, through Nx we get caching too).

We have a `lambda` task in each CDK library with the following configuration: [^1]

```json
  "targets": {
    "lambda": {
      "executor": "@nrwl/esbuild:esbuild",
      "outputs": ["{options.outputPath}"],
      "options": {
        "main": "libs/core/cloud/src/index.ts",
        "tsConfig": "libs/core/cloud/tsconfig.lib.json",
        "outputPath": "dist/lambda/libs/core/cloud/src",
        "skipTypeCheck": true,
        "minify": true,
        "sourcemap": true,
        "bundle": true,
        "target": "node18",
        "format": ["cjs"],
        "esbuildOptions": {
          "outExtension": {
            ".js": ".js"
          }
        },
        "additionalEntryPoints": [
          "libs/core/cloud/src/lib/cloud-front.set-cookies.fn/index.ts",
          "libs/core/cloud/src/lib/cloud-front.signing-key-rotation.fn/index.ts",
          "libs/core/cloud/src/lib/data-neo4j.invoke.fn/index.ts",
          "libs/core/cloud/src/lib/identity-management.custom-message.fn/index.ts",
          "libs/core/cloud/src/lib/identity-management.post-confirmation.fn/index.ts",
          "libs/core/cloud/src/lib/identity-management.pre-sign-up.fn/index.ts"
        ]
      }
    }
  }
```

You can tweak this depending on your target runtime, whether you intend to use commonjs or ecmascript, source maps, minification... `bundle` is however what you want if you want to keep your Lambda all in one file (not that this is mandatory either, but it's rather convenient, and the whole topic of discussion here).

What I don't really like about this configuration, is that we have to explicitly use `additionalEntryPoints` for each Lambda that we bundle. We cannot use glob, and the Nx plugin has a mandatory `main` entry, which we set to the barrel file just for consistency, but is quite a waste actually, since we don't need to bundle that...

Our `deploy` task then depends on `lambda` to make sure we bundle everything before we start deploying.

Finally, we have a construct [^1] (heavily inspired by the NodejsFunction construct [^2]), that will easily reconstruct the full path of the pre-bundled lambda, based on the name of the folder. I guess you already have something similar for Pulumi. In the end, all you need is to reconstruct the right path, but this really depends on what you're using for your deployment.

[^1]: https://gist.github.com/borgoat/718f1650cf6fe7f6fcb6c56a2e783e96
[^2]: https://github.com/aws/aws-cdk/blob/v2.70.0/packages/@aws-cdk/aws-lambda-nodejs/lib/function.ts
